File: 백준/Silver/15666. N과 M （12）/N과 M （12）.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m;
	static int[] numbers;
	static int[] sequence;
	static StringBuilder sb;
	static boolean[] visited;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()); 
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        numbers = new int[n];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++) {
        	numbers[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(numbers);
        
        visited = new boolean[n];
        sequence = new int[m];
        sb = new StringBuilder();
        backtrack(0, 0);
        System.out.println(sb.toString());
    }
   
    private static void backtrack(int depth, int start) {
    	if(depth == m) {
    		for(int num : sequence) {
    			sb.append(num).append(" ");
    		}
    		sb.deleteCharAt(sb.length()-1);
    		sb.append("\n");
    		return;
    	}
    	
    	int lastUsed = -1;
    	for(int i = start; i < n; i++) {
    		if(lastUsed != numbers[i]) {
    			sequence[depth] = numbers[i];
    			lastUsed = numbers[i];
    			backtrack(depth+1, i);
    		}
    	}
    }
    
}

File: 백준/Silver/15666. N과 M （12）/N과 M （12）.kt
package org.example

fun main() {
    val br = System.`in`.bufferedReader()
    val sb = StringBuilder()
    val (N, M) = br.readLine().split(" ").map { it.toInt() }
    val nums = br.readLine().split(" ").map { it.toInt() }.sorted()
    val sequence = IntArray(M)
    fun backtrack(depth: Int, start: Int) {
        if (depth == M) {
            sequence.forEach { sb.append(it).append(" ") }
            sb.deleteCharAt(sb.length - 1).append("\n")
            return
        }
        var lastUsed = -1
        for (i in start until N) {
            if (lastUsed != nums[i]) {
                sequence[depth] = nums[i]
                lastUsed = nums[i]
                backtrack(depth + 1, i)
            }
        }
    }
    backtrack(0, 0)
    print(sb)

    br.close()
}